### PR TITLE
Streaming into Indi detector

### PR DESCRIPTION
**Pull request by iliaplatone**

### CODE REVIEW

**Comment by iliaplatone** [comment]
To add streaming functionalities into a detector driver, you should use the Streamer instance of the StreamManager class, just added into the detector main class.
You also should add the StartStream and StopStream override, and call Streamer->newFrame each time a new frame comes.
FPS have to be fixed and size contains no subframe and height=1

File: libindi/libs/stream/streammanager.cpp [hunk 1]
```diff
@@ -117,15 +83,15 @@ StreamManager::~StreamManager()
...
+    return currentDevice->getDeviceName();
 }
 
 bool StreamManager::initProperties()
 {
     /* Video Stream */
     IUFillSwitch(&StreamS[0], "STREAM_ON", "Stream On", ISS_OFF);
     IUFillSwitch(&StreamS[1], "STREAM_OFF", "Stream Off", ISS_ON);
-    IUFillSwitchVector(&StreamSP, StreamS, NARRAY(StreamS), getDeviceName(), "CCD_VIDEO_STREAM", "Video Stream",
+    IUFillSwitchVector(&StreamSP, StreamS, NARRAY(StreamS), getDeviceName(), "INDI_VIDEO_STREAM", "Video Stream",
```
Comment (knro) [inline libindi/libs/stream/streammanager.cpp]: CCD_VIDEO_STREAM is an INDI standard property. For compatibility reasons, it must be used. Maybe check driver interface and set it to this and for detector define DETECTOR_VIDEO_STREAM or something like that?

File: libindi/libs/stream/streammanager.cpp [hunk 2]
```diff
@@ -279,13 +252,16 @@ void StreamManager::newFrame(const uint8_t * buffer, uint32_t nbytes)
...
     }
-
     std::thread(&StreamManager::asyncStream, this, buffer, nbytes, deltams).detach();
 }
 
 void StreamManager::asyncStream(const uint8_t *buffer, uint32_t nbytes, double deltams)
 {
-    std::unique_lock<std::mutex> guard(currentCCD->ccdBufferLock);
+    if(currentDevice->getDriverInterface() & INDI::DefaultDevice::CCD_INTERFACE) {
+        std::unique_lock<std::mutex> guard(((INDI::CCD*)currentDevice)->ccdBufferLock);
```
Comment (knro) [inline libindi/libs/stream/streammanager.cpp]: use dynamic_cast

File: libindi/libs/stream/streammanager.cpp [hunk 3]
```diff
@@ -294,9 +270,13 @@ void StreamManager::asyncStream(const uint8_t *buffer, uint32_t nbytes, double d
...
         }
-
-        uint32_t npixels = (currentCCD->PrimaryCCD.getSubW() / currentCCD->PrimaryCCD.getBinX()) * (currentCCD->PrimaryCCD.getSubH() / currentCCD->PrimaryCCD.getBinY()) * ((m_PixelFormat == INDI_RGB) ? 3 : 1);
-        //uint32_t npixels = StreamFrameN[CCDChip::FRAME_W].value * StreamFrameN[CCDChip::FRAME_H].value * ((m_PixelFormat == INDI_RGB) ? 3 : 1);
+        uint32_t npixels = 0;
+        if(currentDevice->getDriverInterface() & INDI::DefaultDevice::CCD_INTERFACE) {
+            npixels = (((INDI::CCD*)currentDevice)->PrimaryCCD.getSubW() / ((INDI::CCD*)currentDevice)->PrimaryCCD.getBinX()) * (((INDI::CCD*)currentDevice)->PrimaryCCD.getSubH() / ((INDI::CCD*)currentDevice)->PrimaryCCD.getBinY()) * ((m_PixelFormat == INDI_RGB) ? 3 : 1);
+            //uint32_t npixels = StreamFrameN[CCDChip::FRAME_W].value * StreamFrameN[CCDChip::FRAME_H].value * ((m_PixelFormat == INDI_RGB) ? 3 : 1);
+        } else if(currentDevice->getDriverInterface() & INDI::DefaultDevice::DETECTOR_INTERFACE) {
+            npixels = nbytes * 8 / ((INDI::Detector*)currentDevice)->PrimaryDetector.getBPS();
```
Comment (knro) [inline libindi/libs/stream/streammanager.cpp]: dynamic_cast

File: libindi/libs/stream/streammanager.cpp [hunk 4]
```diff
@@ -598,12 +578,14 @@ bool StreamManager::startRecording()
...
 
-    /* get filter name for pattern substitution */
-    if (currentCCD->CurrentFilterSlot != -1 && currentCCD->CurrentFilterSlot <= static_cast<int>(currentCCD->FilterNames.size()))
-    {
-        filtername      = currentCCD->FilterNames.at(currentCCD->CurrentFilterSlot - 1);
-        patterns["_F_"] = filtername;
-        LOGF_DEBUG("Adding filter pattern %s", filtername.c_str());
+    if(currentDevice->getDriverInterface() & INDI::DefaultDevice::CCD_INTERFACE) {
+        /* get filter name for pattern substitution */
+        if (((INDI::CCD*)currentDevice)->CurrentFilterSlot != -1 && ((INDI::CCD*)currentDevice)->CurrentFilterSlot <= static_cast<int>(((INDI::CCD*)currentDevice)->FilterNames.size()))
```
Comment (knro) [inline libindi/libs/stream/streammanager.cpp]: dynamic_cast

**Comment by iliaplatone** [comment]
After this draft.
I'd propose a streamer interface at this point...
CCDs and light detectors, or even data loggers can use this interface by adding txt, dat, log encoders.
Using it on spectrometers each frame should correspond to a spectra capture.
Light spectrometers could be added and radio receivers should dft subframes or multichannel antennas can be used.
This work needs more time.

**Comment by knro** [comment]
Thanks, looks good!

**Pull request merged**